fmpz_mod_mpoly.h

```diff
@@ -36,364 +36,1073 @@
-                              const char ** x, const fmpz_mod_mpoly_ctx_t ctx);
+/* Scalar operations *********************************************************/
 
-FLINT_DLL void fmpz_mod_mpolyun_print_pretty(const fmpz_mod_mpolyun_t poly,
-                              const char ** x, const fmpz_mod_mpoly_ctx_t ctx);
+FLINT_DLL void fmpz_mod_mpoly_neg(fmpz_mod_mpoly_t A, const fmpz_mod_mpoly_t B,
+                                               const fmpz_mod_mpoly_ctx_t ctx);
 
-/* interp ********************************************************************/
+FLINT_DLL void fmpz_mod_mpoly_scalar_mul_fmpz_mod_invertible(fmpz_mod_mpoly_t A,
+                                    const fmpz_mod_mpoly_t B, const fmpz_t c,
+                                               const fmpz_mod_mpoly_ctx_t ctx);
 
-FLINT_DLL void fmpz_mod_mpolyn_intp_reduce_sm_poly(fmpz_mod_poly_t E,
-                            const fmpz_mod_mpolyn_t A, const fmpz_t alpha,
+FLINT_DLL void fmpz_mod_mpoly_scalar_mul_fmpz(fmpz_mod_mpoly_t A,
+                                    const fmpz_mod_mpoly_t B, const fmpz_t c,
                                                const fmpz_mod_mpoly_ctx_t ctx);
 
-FLINT_DLL void fmpz_mod_mpolyn_intp_lift_sm_poly(fmpz_mod_mpolyn_t A,
-                      const fmpz_mod_poly_t B, const fmpz_mod_mpoly_ctx_t ctx);
+FLINT_DLL void fmpz_mod_mpoly_scalar_mul_ui(fmpz_mod_mpoly_t A,
+                                    const fmpz_mod_mpoly_t B, ulong c,
+                                               const fmpz_mod_mpoly_ctx_t ctx);
 
-FLINT_DLL int fmpz_mod_mpolyn_intp_crt_sm_poly(slong * lastdeg_,
-             fmpz_mod_mpolyn_t F, fmpz_mod_mpolyn_t T, const fmpz_mod_poly_t A,
-                            const fmpz_mod_poly_t modulus, const fmpz_t alpha,
+FLINT_DLL void fmpz_mod_mpoly_scalar_mul_si(fmpz_mod_mpoly_t A,
+                                    const fmpz_mod_mpoly_t B, slong c,
                                                const fmpz_mod_mpoly_ctx_t ctx);
 
-/* helpers *******************************************************************/
+FLINT_DLL void fmpz_mod_mpoly_scalar_addmul_fmpz(fmpz_mod_mpoly_t A,
+                        const fmpz_mod_mpoly_t B, const fmpz_mod_mpoly_t C,
+                               const fmpz_t d, const fmpz_mod_mpoly_ctx_t ctx);
 
-typedef struct {
-    nmod_berlekamp_massey_struct * coeffs;
-    ulong * exps;
-    slong length;
-    slong alloc;
-    slong pointcount;
-} nmod_bma_mpoly_struct;
+FLINT_DLL void fmpz_mod_mpoly_make_monic(fmpz_mod_mpoly_t A,
+                     const fmpz_mod_mpoly_t B, const fmpz_mod_mpoly_ctx_t ctx);
 
-typedef nmod_bma_mpoly_struct nmod_bma_mpoly_t[1];
 
-typedef struct {
-    fmpz_mod_berlekamp_massey_struct * coeffs;
-    ulong * exps;
-    slong length;
-    slong alloc;
-    slong pointcount;
-} fmpz_mod_bma_mpoly_struct;
+/* Differention **************************************************************/
```

Comment:
Differentiation

Reply:
yup. I also found a few more typos and missing functions.

Reply:
Yeah I don't know how much you care about this stuff, but there's still a few files with 2017 and 2018. Grep should find those. There's also a handful of lines wrapping around here and there due to being over 80 chars. I think it is possible to find those with grep and a regexp, though it's not super critical. Also can you check you really intended to delete the files you are. e.g. the fmpq geobuckets, interp.c and various others. Otherwise, looks good and I will merge when you are happy.

Reply:
Also there's some gcd and mpolyu files being deleted, which I assume is intended.

And I think I saw some struct changes. Do we need to modify the Nemo structs?

Reply:
no preexisting structs should have been harmed - do you have a specific example? I am going to change the fmpz_mpoly_struct soon in a later pr when I attempt to reduce code duplication.

Reply:
No just the changes to fmpz_mod_mpoly_ctx_struct and fmpz_mod_mpoly_struct. They are reordered and so on. But probably we have no wrapper for this in Nemo anyway, right?

Reply:
of course